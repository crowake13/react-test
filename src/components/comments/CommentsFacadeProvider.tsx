import React from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { CommentsContext } from '../../stores/comments/comments.context';
import { facade, ICommentsFacade } from '../../stores/comments/comments.facade';
import withRenderLog from '../shared/withRenderLog';

interface CommentsFacadeProviderProps extends Partial<ICommentsFacade> {
  children: any;
}

const CommentsFacadeProvider = ({
  children,
  ...props
}: CommentsFacadeProviderProps) => {
  return (
    <CommentsContext.Provider
      value={{
        errors: props.errors ?? facade.errors,
        error$: props.error$ ?? facade.error$,
        active$: props.active$ ?? facade.active$,
        activate: props.activate ?? facade.activate,
        getAll: props.getAll ?? facade.getAll,
        getById: props.getById ?? facade.getById,
        hashMap: props.hashMap ?? facade.hashMap,
        entitie$: props.entitie$ ?? facade.entitie$,
        isFetchingBySlug: props.isFetchingBySlug ?? facade.isFetchingBySlug,
        isFetching: props.isFetching ?? facade.isFetching,
        isFetching$: props.isFetching$ ?? facade.isFetching$,
        fetch: props.fetch ?? facade.fetch,
        fetchAll: props.fetchAll ?? facade.fetchAll
      }}
    >
      {children}
    </CommentsContext.Provider>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  CommentsFacadeProvider
);
