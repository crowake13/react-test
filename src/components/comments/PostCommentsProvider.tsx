import React, { useState } from 'react';
import * as RENDER_LOG from '../../constants/render-log';
import { useComments } from '../../hooks/comments/comments.hook';
import PostCommentsContext from '../../stores/comments/post-comments.context';
import { ID } from '../../stores/entities/entity.facade';
import withRenderLog from '../shared/withRenderLog';

interface PostCommentsProvider {
  postId: ID;
  areCommentsVisibleOnInit?: boolean;
  children: any;
}

const PostCommentsProvider = ({
  postId,
  areCommentsVisibleOnInit,
  children
}: PostCommentsProvider) => {
  const [areCommentsVisible, toggleCommentsVisibility] = useState(
    areCommentsVisibleOnInit ?? false
  );

  const [, getCommentsByPostId] = useComments();

  return (
    <PostCommentsContext.Provider
      value={{
        comments: getCommentsByPostId(postId),
        areCommentsVisible: areCommentsVisible,
        toggleCommentsVisibility: toggleCommentsVisibility
      }}
    >
      {children}
    </PostCommentsContext.Provider>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  PostCommentsProvider
);
