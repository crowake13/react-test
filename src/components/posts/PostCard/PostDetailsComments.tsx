import React, { useContext } from 'react';
import * as RENDER_LOG from '../../../constants/render-log';
import PostCommentsContext from '../../../stores/comments/post-comments.context';
import { CommentList } from '../../comments/CommentsList';
import withRenderLog from '../../shared/withRenderLog';

interface PostDetailsCommentsProps {
  loadingCommentsLabel: string;
  noCommentsLabel: string;
}

const PostDetailsComments = ({
  loadingCommentsLabel,
  noCommentsLabel
}: PostDetailsCommentsProps) => {
  const { comments } = useContext(PostCommentsContext);

  return (
    <div className="bg-light border-top">
      <div className="card-header">
        {!comments ? (
          loadingCommentsLabel
        ) : !comments.length ? (
          noCommentsLabel
        ) : (
          <span>Comments ({comments.length})</span>
        )}
      </div>
      {comments && comments.length ? <CommentList comments={comments} /> : null}
    </div>
  );
};

export default withRenderLog({ greeting: RENDER_LOG.GREETING })(
  PostDetailsComments
);
