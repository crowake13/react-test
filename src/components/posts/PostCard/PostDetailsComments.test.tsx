import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { Subject } from 'rxjs';
import { Comment } from '../../../stores/comments/comment.model';
import CommentsFacadeProvider from '../../comments/CommentsFacadeProvider';
import PostCommentsProvider from '../../comments/PostCommentsProvider';
import PostDetailsComments from './PostDetailsComments';

let container: HTMLElement | null = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  if (!container) {
    return;
  }

  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const postId = 1;

const comments: Comment[] = [
  {
    postId,
    id: 1,
    name: 'title 1.1',
    email: 'email 1.1',
    body: 'body 1.1'
  },
  {
    postId,
    id: 2,
    name: 'title 1.2',
    email: 'email 1.2',
    body: 'body 1.2'
  }
];

it('renders PostDetailsComments manually', () => {
  const _entitie$ = new Subject<Comment[]>();
  const entitie$ = _entitie$.asObservable();

  const _isFetching$ = new Subject<{ [key: string]: boolean }>();
  const isFetching$ = _isFetching$.asObservable();

  const setIsFetching = (value: boolean, slug: string = 'comments') => {
    _isFetching$.next({
      [slug]: value
    });
  };

  act(() => {
    render(
      <CommentsFacadeProvider entitie$={entitie$} isFetching$={isFetching$}>
        <PostCommentsProvider postId={postId}>
          <PostDetailsComments
            loadingCommentsLabel="Comments are loading..."
            noCommentsLabel="There are no comments"
          />
        </PostCommentsProvider>
      </CommentsFacadeProvider>,
      container
    );
    setIsFetching(false);
  });
  expect(container?.innerHTML).toMatchSnapshot();

  act(() => {
    setIsFetching(true);
  });

  expect(container?.innerHTML).toMatchSnapshot();

  act(() => {
    _entitie$.next(comments);
    setIsFetching(false);
  });

  expect(container?.innerHTML).toMatchSnapshot();
});
