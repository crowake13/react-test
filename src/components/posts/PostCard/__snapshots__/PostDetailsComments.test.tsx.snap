// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PostDetailsComments manually 1`] = `"<div class=\\"bg-light border-top\\"><div class=\\"card-header\\">There are no comments</div></div>"`;

exports[`renders PostDetailsComments manually 2`] = `"<div class=\\"bg-light border-top\\"><div class=\\"card-header\\">Comments are loading...</div></div>"`;

exports[`renders PostDetailsComments manually 3`] = `"<div class=\\"bg-light border-top\\"><div class=\\"card-header\\"><span>Comments (2)</span></div><div class=\\"card-body m-0\\"><div class=\\"mb-2\\"><h6 title=\\"email 1.1\\" class=\\"card-subtitle my-2 text-muted line-clamp-1\\">email 1.1</h6><h5 title=\\"title 1.1\\" class=\\"card-title  line-clamp-1\\">title 1.1</h5><p title=\\"body 1.1\\" class=\\"card-text  line-clamp-1\\">body 1.1</p></div><div class=\\"mb-2 border-top\\"><h6 title=\\"email 1.2\\" class=\\"card-subtitle my-2 text-muted line-clamp-1\\">email 1.2</h6><h5 title=\\"title 1.2\\" class=\\"card-title  line-clamp-1\\">title 1.2</h5><p title=\\"body 1.2\\" class=\\"card-text  line-clamp-1\\">body 1.2</p></div></div></div>"`;
